pkg/registrymirror/store.go

```diff
@@ -0,0 +1,349 @@
+	DeleteImageDigestMirrorSet(uid types.UID) error
+
+	UpsertImageTagMirrorSet(itms *configV1.ImageTagMirrorSet) error
+	DeleteImageTagMirrorSet(uid types.UID) error
+
+	// PullSources will return image references in the order they should be attempted when pulling the image.
+	// The returned slice may (or may not) include the source image based on the registries config.
+	// Source image must include the registry hostname for mirrors to be matched, for example use:
+	// "quay.io/stackrox-io/main:latest" instead of "stackrox-io/main:latest".
+	PullSources(srcImage string) ([]string, error)
+}
+
+// FileStore stores/reads the consolidated registries config from a filesystem.
+type FileStore struct {
+	configPath string
+
+	// holds the mirror sets used for managing registry mirrors.
+	icspRules   map[types.UID]*operatorV1Alpha1.ImageContentSourcePolicy
+	idmsRules   map[types.UID]*configV1.ImageDigestMirrorSet
+	itmsRules   map[types.UID]*configV1.ImageTagMirrorSet
+	ruleRWMutex sync.RWMutex
+
+	systemContext *ciTypes.SystemContext
+
+	updateDelay  time.Duration
+	cancelUpdate concurrency.Signal
+}
+
+// fileStoreOption is used to provide functional options to the fileStore.
+type fileStoreOption func(*FileStore)
+
+// WithConfigPath sets the path to read/write the consolidated registries config.
+func WithConfigPath(path string) fileStoreOption {
+	return func(s *FileStore) { s.configPath = path }
+}
+
+// WithDelay sets a delay before the updated config is written to disk.
+func WithDelay(delay time.Duration) fileStoreOption {
+	return func(s *FileStore) { s.updateDelay = delay }
+}
+
+// NewFileStore creates a new FileStore.
+func NewFileStore(opts ...fileStoreOption) *FileStore {
+	s := &FileStore{
+		configPath:  defaultRegistriesPath,
+		updateDelay: defaultDelay,
+		icspRules:   make(map[types.UID]*operatorV1Alpha1.ImageContentSourcePolicy),
+		idmsRules:   make(map[types.UID]*configV1.ImageDigestMirrorSet),
+		itmsRules:   make(map[types.UID]*configV1.ImageTagMirrorSet),
+	}
+
+	for _, opt := range opts {
+		opt(s)
+	}
+
+	s.systemContext = &ciTypes.SystemContext{
+		SystemRegistriesConfPath: s.configPath,
+	}
+	return s
+}
+
+// Cleanup resets the store which includes in-memory and disk resources.
+func (s *FileStore) Cleanup() {
+	s.ruleRWMutex.Lock()
+	defer s.ruleRWMutex.Unlock()
+
+	s.icspRules = make(map[types.UID]*operatorV1Alpha1.ImageContentSourcePolicy)
+	s.idmsRules = make(map[types.UID]*configV1.ImageDigestMirrorSet)
+	s.itmsRules = make(map[types.UID]*configV1.ImageTagMirrorSet)
+
+	s.cancelUpdate.Signal()
+	if err := os.Remove(s.configPath); err != nil && !os.IsNotExist(err) {
+		log.Warnf("Failed to cleanup registries config at %q: %v", s.configPath, err)
+	}
+	sysregistriesv2.InvalidateCache()
+}
+
+func (s *FileStore) updateConfig() error {
+	if s.updateDelay == 0 {
+		return s.updateConfigNow()
+	}
+
+	return s.updateConfigDelayed()
+}
+
+func (s *FileStore) updateConfigDelayed() error {
+	s.cancelUpdate.Signal()
```

Comment:
this will ensure there is only one goroutine updating the file at any point, right? If not, then we should consider writing to a temp file and then `os.Rename`

Reply:
Thats right, only one goroutine will be updating the file at a given time.